File: web/src/api/Protocol.mjs
import TransportUsb from './TransportUsb.mjs';
import TransportWebSocket from './TransportWebSocket.mjs';
import TransportMidi  from './TransportMidi.mjs';

const LENGTH_OFFSET = 0;
const COMMAND_OFFSET = 2;
const STATUS_OFFSET = 3;
const MSG_HEADER_SIZE = 4;

export default class Protocol {
  constructor(transport, connectionEvent) {
    if ("requestDevice" in transport) {
      this.transport = new TransportUsb(transport, connectionEvent);
    } else if ("requestMIDIAccess" in transport) {
      this.transport = new TransportMidi(transport, connectionEvent);
    } else {
      this.transport = new TransportWebSocket(transport, connectionEvent);
    }
    this.buffer = new Uint8Array();
  }

  get connected() {
    return this.transport.connected;
  }

  connect = reconnect => reconnect ? this.transport.reconnect() : this.transport.connect();
  version = _ => this.transport.version();
  
  async receive() {
    let totalLength = Infinity;
    let msg;
    while (this.buffer.byteLength < totalLength) {
      const rcvData = await this.transport.receive();
      const oldBuffer = this.buffer;
      this.buffer = new Uint8Array(oldBuffer.byteLength + rcvData.byteLength);
      this.buffer.set(oldBuffer);
      this.buffer.set(new Uint8Array(rcvData.buffer), oldBuffer.byteLength);
      msg = new DataView(this.buffer.buffer);
      const length = (this.buffer.byteLength > 2) ? msg.getUint16(LENGTH_OFFSET, true) : Infinity;
      totalLength = MSG_HEADER_SIZE + length;
    }

    const command = msg.getUint8(COMMAND_OFFSET);
    const status = msg.getUint8(STATUS_OFFSET);
    const data = this.buffer.slice(MSG_HEADER_SIZE);
    this.buffer = (this.buffer.byteLength > totalLength) ? this.buffer.slice(totalLength) : new Uint8Array();

    if (status === 0) {
      // console.log('receive', {command, status, data});
      return {command, status, data};
    } else {
      throw new Error(`Response with status ${status}`);
    }
  }

  async sendRequest(command, data = new Uint8Array()) {
    const outBuffer = new Uint8Array(MSG_HEADER_SIZE + data.byteLength);
    const msg = new DataView(outBuffer.buffer);
    msg.setUint16(LENGTH_OFFSET, data.byteLength, true);
    msg.setUint8(COMMAND_OFFSET, command);
    msg.setUint8(STATUS_OFFSET, 0);
    outBuffer.set(data, MSG_HEADER_SIZE);
    await this.transport.send(msg);
    return await this.receive();
  }
}

File: tools/pytocatapedal/src/pytocatapedal/protocol.py
"""Message framing/reassembly, mirroring web/src/api/Protocol.mjs.

Every message on the wire is a 4-byte header (uint16 LE length, uint8
command, uint8 status) followed by `length` payload bytes. Unlike the JS
version -- which serializes concurrent callers through an internal Promise
queue -- this is a synchronous, blocking port: a single lock around
send+receive is enough since callers issue one request at a time.
"""
import struct as _struct
import threading
from typing import Tuple

LENGTH_OFFSET = 0
COMMAND_OFFSET = 2
STATUS_OFFSET = 3
MSG_HEADER_SIZE = 4


class Protocol:
    def __init__(self, transport, connection_event=None):
        self.transport = transport
        if connection_event is not None:
            self.transport.connection_event = connection_event
        self._buffer = bytearray()
        self._lock = threading.Lock()

    @property
    def connected(self) -> bool:
        return self.transport.connected

    def connect(self, reconnect: bool = False):
        return self.transport.reconnect() if reconnect else self.transport.connect()

    def version(self):
        return self.transport.version()

    def _receive(self) -> Tuple[int, bytes]:
        total_length = None
        while total_length is None or len(self._buffer) < total_length:
            self._buffer += self.transport.receive()
            if len(self._buffer) > 2:
                length = _struct.unpack_from("<H", self._buffer, LENGTH_OFFSET)[0]
                total_length = MSG_HEADER_SIZE + length

        command = self._buffer[COMMAND_OFFSET]
        status = self._buffer[STATUS_OFFSET]
        # Mirrors the JS side: this is every trailing buffered byte, not just
        # `length` of them -- harmless, since struct parsing only consumes as
        # many bytes as its scheme declares.
        data = bytes(self._buffer[MSG_HEADER_SIZE:])
        self._buffer = self._buffer[total_length:] if len(self._buffer) > total_length else bytearray()

        if status != 0:
            raise RuntimeError(f"Response with status {status}")
        return command, data

    def send_request(self, command: int, data: bytes = b"") -> Tuple[int, bytes]:
        with self._lock:
            header = _struct.pack("<HBB", len(data), command, 0)
            self.transport.send(header + bytes(data))
            return self._receive()
